### PR TITLE
Fix(Draggable): compensate for container CSS scale

### LINKED ISSUE #6001: Dragging works wrong on CSS3 scaled map
## How to reproduce

- Leaflet version I'm using: current master
- Browser (with version) I'm using: Chrome 63.0.3239.84
- OS/Platform (with version) I'm using: Win 10

## What behaviour I'm expecting and which behaviour I'm seeing

Expected that cursor should not change position relative to object, that you are dragging.

But if map is scaled by CSS3 transform then while you are dragging, cursor shall be more and more far away from position, there you clicked to start dragging. 

Markers also affected by that bug. While you are dragging marker position of it shall differ from the cursor's one.

## Minimal example reproducing the issue

Try to drag map or marker on this page, and you'll see bugs, described above.

https://leaflet-scaling-reproduce.herokuapp.com/old.html

## Fix

#5997

### LINKED ISSUE #6002: Mouse events coordinates are wrong on CSS3 scaled map
## How to reproduce

- Leaflet version I'm using: current master
- Browser (with version) I'm using: Chrome 63.0.3239.84
- OS/Platform (with version) I'm using: Win 10

## What behaviour I'm expecting and which behaviour I'm seeing

Expected that coordinates of mouse click or wheel scrolling shall be the same as actual pointer position. 

But if map is scaled by CSS3 transform they are different.

## Minimal example reproducing the issue

Try to click the map

https://leaflet-scaling-reproduce.herokuapp.com/old.html

## Fix

#5997

### PR DESCRIPTION
**Pull request by ghybs**
Closes #5997 as replacement PR.
Fixes #6001, fixes #6002.

As done by #5997, this PR patches `Draggable` class so that it measures the user drag through `DomEvent.getMousePosition` instead of directly reading the `event.clientX` / `event.clientY` values, and we get a chance to compensate for the container CSS scale.

In order to avoid having to explicitly cache some map container, `Draggable` automatically looks for the `_element`'s closest parent node with a non null size, using a new `DomUtil.getSizedParentNode` function. This way, the scale computation can determine the value that applies on the `container` (an edge case is when some skipped intermediate parent nodes - with null size - also apply a CSS scale).

Added 2 dedicated test suites for the Map Drag and Marker Drag.
Also modified the `map-scaled.html` debug page to test both features more easily.

### CODE REVIEW

**Comment by ghybs** [comment]
Also closes #6001 and closes #6002

File: src/dom/Draggable.js [hunk 1]
```diff
@@ -112,9 +111,12 @@ export var Draggable = Evented.extend({
 		// Fired when a drag is about to start.
 		this.fire('down');
 
-		var first = e.touches ? e.touches[0] : e;
+		var first = e.touches ? e.touches[0] : e,
+		    sizedParent = DomUtil.getSizedParentNode(this._element);
```
Comment (cherniavskii) [inline src/dom/Draggable.js]: Sorry for taking so long to review this PR :)

Do we need to look for sized parent node *every* time `_onDown` or `_onMove` are fired?
Unless I'm mistaken, it's enough to do that once in `initialize` method

Comment (ghybs) [reply src/dom/Draggable.js]: Hi,

Thank you for your review!

Well, there is the edge case where container scale is changed dynamically…

A more common case is when the map container has an initially null size (therefore leading to usage of `invalidateSize` later on).
Since `Draggable` is initialized on the map initialization, it would not be able to compute the scaled size at that moment.

Would be happy to find a suitable way to cache the scale though indeed, but covering as much use cases as possible and providing a workaround for other cases…

Comment (ghybs) [reply src/dom/Draggable.js]: Maybe cache on `_onDown`, hoping that the scale would not change for the drag duration?

Comment (cherniavskii) [reply src/dom/Draggable.js]: >Well, there is the edge case where container scale is changed dynamically…

I don't think it's a problem - we will store only *reference* to that element and we'll get it's size on each `_onDown` and `_onMove`

Comment (cherniavskii) [reply src/dom/Draggable.js]: >A more common case is when the map container has an initially null size (therefore leading to usage of invalidateSize later on).

Huh, that may be a problem

>Since Draggable is initialized on the map initialization, it would not be able to compute the scaled size at that moment.

The real problem is that in case of null size map container it will skip it because of null size :/

Comment (cherniavskii) [reply src/dom/Draggable.js]: What if we replace `getSizedParentNode` with `getMapContainer` function (like [that](https://github.com/Leaflet/Leaflet/pull/5944/files#diff-b2be552947d40375eca631cc4a3cd4ddR44))?
This way we can store reference to map container during `initialize` method and read it's size on each `onDown/onMove`

Comment (ghybs) [reply src/dom/Draggable.js]: > we will store only _reference_ to that element and we'll get it's size on each `_onDown` and `_onMove`

True, looks like I already forgot how this PR works… :sweat_smile:

> What if we replace `getSizedParentNode` with `getMapContainer` function?

That is interesting indeed. The advantage is that the container detection is only done once, at instantiation. But now `Draggable` class becomes aware of map characteristics, whereas until now [it can be used independently from any map](https://github.com/Leaflet/Leaflet/pull/5997#discussion_r167515814). Same for `DomUtil`.

I would tend to keep `getSizedParentNode` (to keep `Draggable` agnostic) and to cache the found container in `_onDown`. That would avoid repeatedly looking for that container in `_onMove`. We can reasonably assume that if a drag is started, then there is a non null size parent container. If its scale changes dynamically, it will be accounted for in `getMousePosition` (but a further optimization could also cache the scale in `_onDown`, assuming it will not change during the drag, or user will restart their drag anyway due to the disruption in the map display when its scale changes).

IMHO I do not think looking for the container on every `_onDown` would be such a performance issue: it is just walking up the DOM parent hierarchy. And `_onDown` is triggered on user drag start, much less often than `_onMove`.

What do you think?

Comment (cherniavskii) [reply src/dom/Draggable.js]: > But now Draggable class becomes aware of map characteristics, whereas until now it can be used independently from any map.

You are right, let's keep `Draggable` agnostic!

So we have two options:
- cache sized container
- cache scale

My vote is for caching scale, but I don't really think there's significant performance difference between those approaches.
The important thing is that either way `_onMove` will be much faster :)

File: src/dom/DomUtil.js [hunk 2]
```diff
@@ -319,3 +319,12 @@ export function restoreOutline() {
 	_outlineStyle = undefined;
 	DomEvent.off(window, 'keydown', restoreOutline);
 }
+
+// @function getSizedParentNode(el: HTMLElement)
```
Comment (cherniavskii) [inline src/dom/DomUtil.js]: looks like you forgot to document return value :)

File: src/dom/Draggable.js (hunk 1 above)
Comment (cherniavskii) [reply src/dom/Draggable.js]: Thanks for sharing that link, didn't know about that!

But in Appendix section it says:
> Reflow only has a cost if the document has changed and invalidated the style or layout

So I'm wondering if reflow has a cost in our case

Comment (cherniavskii) [reply src/dom/Draggable.js]: @ghybs I agree ;)

**Comment by ghybs** [comment]
Refactored.

Not super happy with `DomUtil.getScale` being used in `DomEvent.getMousePosition`, leading to a new circular dependency…

This shows that atomizing modules by function (as done in d3), at least for utilities that do not depend on `L.Class`, would better align with the spirit of ES2015 modules.
